firewood/src/proofs/diff.rs

```diff
@@ -0,0 +1,2028 @@
+/// case that the left trie node has the same path and hash as the right trie node. We
+/// can then skip their children in the traversal. `prev_num_children` is an u8 which
+/// is big enough since there can at most be 16 children per node.
+///
+/// Note that this skipping the children functionality can only be called once after a
+/// call to `next`. We cannot call `skip_children` twice in order to skip all of the
+/// children from the current node's parent. This functionality is not necessary for
+/// our use case as a hash match on the parent would have been determined previously
+/// since we are performing in-order traversal.
+struct PreOrderIterator<'a> {
+    trie: &'a dyn TrieReader,
+    stack: Vec<TraversalStackFrame>,
+    prev_num_children: u8,
+}
+
+/// Ignoring possible arithmetic side effects since we are only incrementing
+/// `prev_num_children` once per child and there can only be at most 16 children per
+/// node and `prev_num_children` is an u8.
+#[allow(clippy::arithmetic_side_effects)]
+impl PreOrderIterator<'_> {
+    // Creates a NodeState from a pre-path and Child that were popped off the
+    // traversal stack. The NodeState is used to compare nodes between tries.
+    fn create_node_state(
+        &self,
+        pre_path: &Path,
+        node: Child,
+        mut node_hash: Option<TrieHash>,
+    ) -> Result<NodeState, FileIoError> {
+        #[cfg(not(feature = "ethhash"))]
+        let node = match node {
+            Child::Node(node) => node.clone().into(),
+            Child::AddressWithHash(addr, hash) => {
+                node_hash = Some(hash);
+                self.trie.read_node(addr)?
+            }
+            Child::MaybePersisted(maybe_persisted, hash) => {
+                node_hash = Some(hash);
+                maybe_persisted.as_shared_node(&self.trie)?
+            }
+        };
+        #[cfg(feature = "ethhash")]
+        let node = match node {
+            Child::Node(node) => node.clone().into(),
+            Child::AddressWithHash(addr, hash) => {
+                node_hash = Some(hash.into());
+                self.trie.read_node(addr)?
+            }
+            Child::MaybePersisted(maybe_persisted, hash) => {
+                node_hash = Some(hash.into());
+                maybe_persisted.as_shared_node(&self.trie)?
+            }
+        };
+
+        // Compute the full path for the node. This is used to compare with the full path
+        // of the current node from the other trie.
+        let full_path = Path::from_nibbles_iterator(
+            pre_path
+                .iter()
+                .copied()
+                .chain(node.partial_path().iter().copied()),
+        );
+        Ok(NodeState {
+            path: full_path,
+            node,
+            hash: node_hash,
+        })
+    }
+
+    fn next(&mut self) -> Result<Option<NodeState>, FileIoError> {
+        counter!("firewood.diff_merkle", "traversal" => "next").increment(1);
```

Comment:
This doesn't properly register the metric. Use firewood_counter! here instead.

Also, let's change this to "firewood.change_proof.diff_traversal" or, even better, "firewood.change_proof.next"